### PR TITLE
UI: move onroad hud to Qt

### PR DESCRIPTION
**Pull request by deanlee**
split from https://github.com/commaai/openpilot/pull/22173
hud2

### CODE REVIEW

File: selfdrive/ui/qt/onroad.cc [hunk 1]
```diff
@@ -162,6 +172,108 @@ void OnroadAlerts::paintEvent(QPaintEvent *event) {
   }
 }
 
+// OnroadHud
+OnroadHud::OnroadHud(QWidget *parent) : QWidget(parent) {
+  engage_img = QPixmap("../assets/img_chffr_wheel.png").scaled(img_size, img_size, Qt::KeepAspectRatio, Qt::SmoothTransformation);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/onroad.cc:177]: let's make a helper for this in qt utils. we do this in quite a few places

Comment (adeebshihadeh) [reply selfdrive/ui/qt/onroad.cc:177]: think it's best if the helper and using it is its own PR

Comment (deanlee) [reply selfdrive/ui/qt/onroad.cc:177]: i was thinking about this :D

**Comment by adeebshihadeh** [comment]
Can you post some image diffs for verification with these UI refactors?

**Comment by deanlee** [comment]
the "red color" on the background is painted by nanovg at the same time with COLOR_RED_ALPHA:

head3

they are almost the same, but it’s hard to be exactly the same because nanovg scales the font on painting.

**Comment by deanlee** [comment]
improved the boundrect:
h1

old hud(nanovg) paint at the same time with Qt hud, no diff now. 

good

File: selfdrive/ui/paint.cc [hunk 2]
```diff
@@ -213,11 +142,8 @@ static void ui_draw_vision(UIState *s) {
   if (scene->world_objects_visible) {
     ui_draw_world(s);
   }
-  // Set Speed, Current Speed, Status/Events
+  // TODO: do we need this?
```
Comment (adeebshihadeh) [inline selfdrive/ui/paint.cc:145]: yes, we still need this. let's move it into it's own widget

File: selfdrive/ui/qt/onroad.cc [hunk 3]
```diff
@@ -27,11 +27,16 @@ OnroadWindow::OnroadWindow(QWidget *parent) : QWidget(parent) {
 
   stacked_layout->addWidget(split_wrapper);
 
+  hud = new OnroadHud(this);
+  hud->setAttribute(Qt::WA_TransparentForMouseEvents, true);
+  stacked_layout->addWidget(hud);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/onroad.cc:32]: this should be part of a new stacked layout together with the nvg widget, then you can get rid of setMapWidth

File: selfdrive/ui/qt/onroad.cc [hunk 4]
```diff
@@ -150,6 +159,99 @@ void OnroadAlerts::paintEvent(QPaintEvent *event) {
...
+// OnroadHud
+OnroadHud::OnroadHud(QWidget *parent) : QWidget(parent) {
+  engage_img = QPixmap("../assets/img_chffr_wheel.png").scaled(img_size, img_size, Qt::KeepAspectRatio, Qt::SmoothTransformation);
+  dm_img = QPixmap("../assets/img_driver_face.png").scaled(img_size, img_size, Qt::KeepAspectRatio, Qt::SmoothTransformation);
+
+  connect(this, &OnroadHud::valueChanged, [=] { update(); });
+}
+
+void OnroadHud::updateState(const UIState &s) {
+  auto getMaxSpeed = [&s](float maxspeed) {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/onroad.cc:171]: no need for the lambda here

File: selfdrive/ui/qt/onroad.cc [hunk 5]
```diff
@@ -17,13 +17,18 @@ OnroadWindow::OnroadWindow(QWidget *parent) : QWidget(parent) {
   stacked_layout->setStackingMode(QStackedLayout::StackAll);
   main_layout->addLayout(stacked_layout);
 
+  QStackedLayout *hud_stacked_layout = new QStackedLayout;
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/onroad.cc:20]: maybe call this road_view_layout?

**Pull request merged**